File: api/routers/b2c.py
from fastapi import APIRouter, HTTPException, Request
from api.routers.websocket import broadcast_payment_status
import httpx
from api.models.b2c_schemas import (
    B2CPaymentRequest,
    B2CPaymentResponse,
    B2CCallback,
    SuccessResponse,
)
from api.services.auth_service import AuthService
from api.core.config import settings
from api.core.utils import format_phone_number
from Crypto.PublicKey import RSA
from Crypto.Cipher import PKCS1_v1_5
import base64

router = APIRouter()


def encrypt_initiator_password(password: str) -> str:
    """

    Args:
        password: Plain text initiator password

    Returns:
        Base64 encoded encrypted password
    """
    try:

        # Example implementation (uncomment when you have the certificate):
        with open('certificates/ProductionCertificate.cer', 'rb') as cert_file:
            cert_data = cert_file.read()
            public_key = RSA.importKey(cert_data)
            cipher = PKCS1_v1_5.new(public_key)
            encrypted = cipher.encrypt(password.encode())
            return base64.b64encode(encrypted).decode()



    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Error encrypting password: {str(e)}"
        )


@router.post("/payment", response_model=B2CPaymentResponse)
async def initiate_b2c_payment(request: B2CPaymentRequest):
    """
    Initiate B2C Payment (Business to Customer)

    This endpoint sends money from your business to a customer's M-Pesa account.
    Use cases:
    - Salary payments
    - Cashback/refunds
    - Promotional payments
    - Winnings/prizes
    - Loan disbursements

    CommandID options:
    - SalaryPayment: Supports both registered and unregistered customers
    - BusinessPayment: Normal business payment (registered customers only)
    - PromotionPayment: Promotional payment with congratulatory message (registered only)
    """
    try:
        # Get access token
        access_token = await AuthService.get_access_token()

        # Format phone number
        phone_number = format_phone_number(request.phone_number)

        # Encrypt initiator password
        security_credential = encrypt_initiator_password(settings.INITIATOR_PASSWORD)

        # Prepare request payload
        payload = {
            "OriginatorConversationID": request.originator_conversation_id,
            "InitiatorName": settings.INITIATOR_NAME,
            "SecurityCredential": security_credential,
            "CommandID": request.command_id,
            "Amount": request.amount,
            "PartyA": settings.BUSINESS_SHORT_CODE,
            "PartyB": phone_number,
            "Remarks": request.remarks,
            "QueueTimeOutURL": settings.B2C_TIMEOUT_URL,
            "ResultURL": settings.B2C_RESULT_URL,
            "Occasion": request.occasion or ""
        }

        # Make API request
        headers = {
            "Authorization": f"Bearer {access_token}",
            "Content-Type": "application/json"
        }

        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{settings.BASE_URL}/mpesa/b2c/v3/paymentrequest",
                json=payload,
                headers=headers,
                timeout=settings.API_TIMEOUT
            )

            if response.status_code != 200:
                error_data = response.json() if response.text else {}
                raise HTTPException(
                    status_code=response.status_code,
                    detail={
                        "error": "B2C payment request failed",
                        "error_code": error_data.get("errorCode", "UNKNOWN"),
                        "error_message": error_data.get("errorMessage", response.text),
                        "request_id": error_data.get("requestId")
                    }
                )

            data = response.json()

            return B2CPaymentResponse(
                conversation_id=data.get("ConversationID"),
                originator_conversation_id=data.get("OriginatorConversationID"),
                response_code=data.get("ResponseCode"),
                response_description=data.get("ResponseDescription")
            )

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Error initiating B2C payment: {str(e)}"
        )


@router.post("/result", response_model=SuccessResponse)
async def b2c_result_callback(request: Request):
    """
    Result callback endpoint for B2C payments.
    M-Pesa sends transaction results here after processing.
    """
    try:
        # Get the callback data
        body = await request.json()

        # Extract result details
        result = body.get("Result", {})

        callback_data = B2CCallback(
            conversation_id=result.get("ConversationID"),
            originator_conversation_id=result.get("OriginatorConversationID"),
            transaction_id=result.get("TransactionID"),
            result_code=result.get("ResultCode"),
            result_desc=result.get("ResultDesc"),
            result_type=result.get("ResultType"),
            result_parameters=result.get("ResultParameters"),
            reference_data=result.get("ReferenceData")
        )

        # Broadcast the result to frontend via WebSocket
        await broadcast_payment_status({
            "type": "b2c_result",
            "data": callback_data.model_dump()
        })

        # Process based on result code
        if callback_data.result_code == 0:
            # Successful transaction
            print(f"✓ B2C Payment successful: {callback_data.transaction_id}")

            # Extract transaction details
            if callback_data.result_parameters:
                items = callback_data.result_parameters.get("ResultParameter", [])
                transaction_details = {item["Key"]: item.get("Value") for item in items}

                print(f"Transaction Details:")
                print(f"  - Amount: {transaction_details.get('TransactionAmount')}")
                print(f"  - Receipt: {transaction_details.get('TransactionReceipt')}")
                print(f"  - Recipient: {transaction_details.get('ReceiverPartyPublicName')}")
                print(f"  - Completed: {transaction_details.get('TransactionCompletedDateTime')}")
                print(f"  - Working Account Balance: {transaction_details.get('B2CWorkingAccountAvailableFunds')}")

                # Here you would typically:
                # 1. Update your database with transaction details
                # 2. Send confirmation notification to customer
                # 3. Update accounting records
                # 4. Trigger any post-payment business logic

        else:
            # Failed transaction
            print(f"✗ B2C Payment failed: {callback_data.result_desc}")
            print(f"  Result Code: {callback_data.result_code}")

            # Handle specific error codes
            if callback_data.result_code == 2001:
                print("  Error: Invalid initiator information")

            # Log to database or monitoring system

        # Return success to M-Pesa
        return SuccessResponse(
            message="Result callback received successfully",
            data=callback_data.model_dump()
        )

    except Exception as e:
        # Log error but return success to M-Pesa to avoid retries
        print(f"Error processing B2C result callback: {str(e)}")
        return SuccessResponse(
            message="Callback received",
            data={"error": str(e)}
        )


@router.post("/timeout", response_model=SuccessResponse)
async def b2c_timeout_callback(request: Request):
    """
    Timeout callback endpoint for B2C payments.
    M-Pesa sends notification here if payment request times out in queue.
    """
    try:
        # Get the callback data
        body = await request.json()

        print(f"⚠ B2C Payment timeout received")
        print(f"Timeout data: {body}")

        # Broadcast timeout to frontend
        await broadcast_payment_status({
            "type": "b2c_timeout",
            "data": body
        })

        # Handle timeout scenario
        # 1. Update database status
        # 2. Notify user
        # 3. Consider retry logic

        return SuccessResponse(
            message="Timeout callback received successfully",
            data=body
        )

    except Exception as e:
        print(f"Error processing B2C timeout callback: {str(e)}")
        return SuccessResponse(
            message="Callback received",
            data={"error": str(e)}
        )

File: b2c.js
// Configuration
const API_BASE_URL = "https://mpesabeginner.vercel.app/api/v1/b2c";
const WS_URL = "wss://mpesabeginner.vercel.app/ws/payments";

// Store current payment request
let currentPayment = {
    conversation_id: null,
    originator_conversation_id: null,
};

// WebSocket connection
let ws = null;
let reconnectInterval = null;

// DOM Elements
const form = document.getElementById("b2cForm");
const phoneInput = document.getElementById("phoneNumber");
const amountInput = document.getElementById("amount");
const commandSelect = document.getElementById("commandId");
const remarksInput = document.getElementById("remarks");
const occasionInput = document.getElementById("occasion");
const submitBtn = document.getElementById("submitBtn");
const responseBox = document.getElementById("responseBox");
const wsStatus = document.getElementById("wsStatus");
const commandInfo = document.getElementById("commandInfo");

// Command ID descriptions
const commandDescriptions = {
    BusinessPayment: "Standard business payment for registered M-Pesa customers",
    SalaryPayment: "Salary payment - supports both registered and unregistered M-Pesa users",
    PromotionPayment: "Promotional payment with congratulatory message (registered only)"
};

// ==================== WebSocket Functions ====================

function connectWebSocket() {
    try {
        ws = new WebSocket(WS_URL);

        ws.onopen = () => {
            console.log("✅ WebSocket connected");
            updateWSStatus(true);
            clearInterval(reconnectInterval);
        };

        ws.onmessage = (event) => {
            try {
                const data = JSON.parse(event.data);
                console.log("📨 WebSocket message received:", data);

                // Check if this is a B2C result
                if (data.type === "b2c_result") {
                    handleB2CCallback(data.data);
                } else if (data.type === "b2c_timeout") {
                    handleB2CTimeout(data.data);
                } else {
                    // Check if it matches our current payment
                    if (
                        data.conversation_id === currentPayment.conversation_id ||
                        data.originator_conversation_id === currentPayment.originator_conversation_id
                    ) {
                        handleB2CCallback(data);
                    }
                }
            } catch (error) {
                console.error("Error parsing WebSocket message:", error);
            }
        };

        ws.onerror = (error) => {
            console.error("❌ WebSocket error:", error);
            updateWSStatus(false);
        };

        ws.onclose = () => {
            console.log("🔴 WebSocket disconnected");
            updateWSStatus(false);
            // Attempt to reconnect after 3 seconds
            reconnectInterval = setInterval(() => {
                console.log("🔄 Attempting to reconnect...");
                connectWebSocket();
            }, 3000);
        };
    } catch (error) {
        console.error("Failed to connect WebSocket:", error);
        updateWSStatus(false);
    }
}

function updateWSStatus(connected) {
    if (connected) {
        wsStatus.textContent = "🟢 Connected";
        wsStatus.className = "ws-status ws-connected";
    } else {
        wsStatus.textContent = "🔴 Disconnected";
        wsStatus.className = "ws-status ws-disconnected";
    }
}

// ==================== Utility Functions ====================

function generateTimestamp() {
    return new Date().toISOString();
}

function generateUniqueId() {
    return `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
}

function formatPhoneNumber(phone) {
    // Remove spaces, dashes, and plus sign
    phone = phone.replace(/[\s\-+]/g, "");
    
    // Convert 07XX to 2547XX
    if (phone.startsWith("0")) {
        phone = "254" + phone.substring(1);
    }
    
    return phone;
}

function appendToResponseBox(message, status = null) {
    const timestamp = new Date().toLocaleTimeString();
    
    if (status) {
        const badge = document.createElement("span");
        badge.className = `status-badge status-${status}`;
        badge.textContent = status.toUpperCase();
        responseBox.innerHTML = "";
        responseBox.appendChild(badge);
        responseBox.appendChild(document.createTextNode("\n"));
    }
    
    responseBox.textContent += `[${timestamp}] ${message}\n`;
    responseBox.scrollTop = responseBox.scrollHeight;
}

function clearResponseBox() {
    responseBox.innerHTML = '<span style="color: #999;">Processing payment...</span>';
}

// ==================== B2C Payment Functions ====================

async function initiateB2CPayment(payload) {
    const timestamp = generateTimestamp();
    
    try {
        const response = await fetch(`${API_BASE_URL}/payment`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                "X-Request-Timestamp": timestamp,
                "X-Merchant-Key": "merchant_123", // Replace with your actual key
            },
            body: JSON.stringify(payload),
        });

        const data = await response.json();

        if (!response.ok) {
            throw new Error(
                data.detail?.error_message || 
                JSON.stringify(data.detail) || 
                "Payment request failed"
            );
        }

        return data;
    } catch (error) {
        throw error;
    }
}

function handleB2CCallback(callbackData) {
    console.log("Processing B2C callback:", callbackData);
    
    // Clear previous content
    responseBox.innerHTML = "";
    
    if (callbackData.result_code === 0) {
        // Successful payment
        appendToResponseBox("✅ Payment Successful!", "success");
        appendToResponseBox(`Transaction ID: ${callbackData.transaction_id}`);
        
        // Extract and display transaction details
        if (callbackData.result_parameters) {
            const items = callbackData.result_parameters.ResultParameter || [];
            const details = {};
            items.forEach(item => {
                details[item.Key] = item.Value;
            });
            
            appendToResponseBox("\n📊 Transaction Details:");
            appendToResponseBox(`Amount: KES ${details.TransactionAmount}`);
            appendToResponseBox(`Receipt: ${details.TransactionReceipt}`);
            appendToResponseBox(`Recipient: ${details.ReceiverPartyPublicName}`);
            appendToResponseBox(`Completed: ${details.TransactionCompletedDateTime}`);
            
            if (details.B2CRecipientIsRegisteredCustomer) {
                const isRegistered = details.B2CRecipientIsRegisteredCustomer === "Y";
                appendToResponseBox(
                    `Customer Status: ${isRegistered ? "Registered" : "Unregistered"}`
                );
            }
            
            appendToResponseBox("\n💰 Account Balances:");
            appendToResponseBox(
                `Working Account: KES ${details.B2CWorkingAccountAvailableFunds}`
            );
            appendToResponseBox(
                `Utility Account: KES ${details.B2CUtilityAccountAvailableFunds}`
            );
        }
    } else {
        // Failed payment
        appendToResponseBox("❌ Payment Failed!", "error");
        appendToResponseBox(`Error Code: ${callbackData.result_code}`);
        appendToResponseBox(`Description: ${callbackData.result_desc}`);
        
        // Handle common error codes
        if (callbackData.result_code === 2001) {
            appendToResponseBox("\n⚠️ Initiator information is invalid");
            appendToResponseBox("Check your INITIATOR_NAME and SECURITY_CREDENTIAL");
        } else if (callbackData.result_code === 1) {
            appendToResponseBox("\n⚠️ Insufficient balance in account");
        }
    }
    
    // Re-enable submit button
    submitBtn.disabled = false;
    submitBtn.textContent = "Send Payment";
    submitBtn.classList.remove("loading");
}

function handleB2CTimeout(timeoutData) {
    console.log("B2C payment timeout:", timeoutData);
    
    responseBox.innerHTML = "";
    appendToResponseBox("⏱️ Payment Request Timeout", "error");
    appendToResponseBox("The payment request took too long to process.");
    appendToResponseBox("Please try again or contact support.");
    
    // Re-enable submit button
    submitBtn.disabled = false;
    submitBtn.textContent = "Send Payment";
    submitBtn.classList.remove("loading");
}

// ==================== Event Listeners ====================

// Update command info when selection changes
commandSelect.addEventListener("change", (e) => {
    commandInfo.textContent = commandDescriptions[e.target.value];
});

// Form submission
form.addEventListener("submit", async (e) => {
    e.preventDefault();
    
    // Disable submit button
    submitBtn.disabled = true;
    submitBtn.textContent = "Processing...";
    submitBtn.classList.add("loading");
    
    // Clear response box
    clearResponseBox();
    
    // Prepare payload
    const payload = {
        phone_number: formatPhoneNumber(phoneInput.value.trim()),
        amount: parseInt(amountInput.value),
        command_id: commandSelect.value,
        remarks: remarksInput.value.trim() || "Payment",
        occasion: occasionInput.value.trim() || null,
        originator_conversation_id: generateUniqueId(),
    };
    
    appendToResponseBox("📤 Initiating B2C payment...", "processing");
    appendToResponseBox(`Phone: ${payload.phone_number}`);
    appendToResponseBox(`Amount: KES ${payload.amount}`);
    appendToResponseBox(`Type: ${payload.command_id}`);
    
    try {
        const response = await initiateB2CPayment(payload);
        
        // Store payment identifiers
        currentPayment.conversation_id = response.conversation_id;
        currentPayment.originator_conversation_id = response.originator_conversation_id;
        
        // Display initial response
        responseBox.innerHTML = "";
        appendToResponseBox("✅ Payment Request Accepted", "pending");
        appendToResponseBox(`Conversation ID: ${response.conversation_id}`);
        appendToResponseBox(`Status: ${response.response_description}`);
        appendToResponseBox("\n⏳ Waiting for payment confirmation...");
        appendToResponseBox("This may take a few moments.");
        
        // Note: The actual result will come via WebSocket callback
        // Button will be re-enabled when callback is received
        
    } catch (error) {
        responseBox.innerHTML = "";
        appendToResponseBox("❌ Payment Request Failed", "error");
        appendToResponseBox(`Error: ${error.message}`);
        
        // Re-enable submit button
        submitBtn.disabled = false;
        submitBtn.textContent = "Send Payment";
        submitBtn.classList.remove("loading");
    }
});

// Phone number formatting on input
phoneInput.addEventListener("blur", (e) => {
    try {
        const formatted = formatPhoneNumber(e.target.value);
        e.target.value = formatted;
    } catch (error) {
        // Invalid format, let validation handle it
    }
});

// ==================== Initialize ====================

document.addEventListener("DOMContentLoaded", () => {
    console.log("🚀 B2C Payment Interface Loaded");
    console.log(`API: ${API_BASE_URL}`);
    console.log(`WebSocket: ${WS_URL}`);
    
    // Connect WebSocket
    connectWebSocket();
    
    // Set default command info
    commandInfo.textContent = commandDescriptions[commandSelect.value];
});

// Cleanup on page unload
window.addEventListener("beforeunload", () => {
    if (ws) {
        ws.close();
    }
    if (reconnectInterval) {
        clearInterval(reconnectInterval);
    }
});